File: CPP/108.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* sortedArrayToBST(vector<int>& nums) {
        return balancer(nums, 0, nums.size() - 1);
    }

    TreeNode* balancer(vector<int>& nums, int start, int end){
        if (start <= end){ // Not empty
            int mid = start + (end - start) / 2;
            TreeNode *node = new TreeNode (nums[mid]);
            node->left = balancer(nums, start, mid - 1);
            node->right = balancer(nums, mid + 1, end);
            return node;
        }

        return nullptr;
    }
};

File: Python/108.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def balancer(self, nums: List[int], start: int, end: int) -> TreeNode:
        if start <= end:
            mid = start + (end - start) // 2;

            node = TreeNode(nums[mid])
            node.left = self.balancer(nums, start, mid -1);
            node.right = self.balancer(nums, mid + 1, end);
            return node;
        else:
            return None

    def sortedArrayToBST(self, nums: List[int]) -> TreeNode:
        return self.balancer(nums, 0, len(nums) - 1);
